prog/dftb+/lib_math/lapackroutines.F90

```diff
@@ -1155,6 +1157,111 @@ subroutine larnv_dblecplx(iDist,iSeed,x)
   end subroutine larnv_dblecplx
 
 
+  !> real svd decomposition of matrix A into left and right vectors and singular values
+  subroutine sgesvd_real(A,u,sigma,vt)
+
+    !> matrix to decompose, warning the matrix is over-written by the routine
+    real(rsp), intent(inout) :: A(:,:)
+
+    !> first min(m,n) columns of u hold the left singular vector on return
+    real(rsp), intent(out) :: u(:,:)
+
+    !> holds the singular values on return
+    real(rsp), intent(out) :: sigma(:)
+
+    !> first min(m,n) columns of vt hold the right singular vector on return - warning this matrix
+    !> is returned transpose(conjugated()) i.e. A = u.s.vt and all non-returned singular vectors are
+    !> zero!
+    real(rsp), intent(out) :: vt(:,:)
+
+    integer :: n, m, mn, lda, lwork, ldu, ldvt, info
+    real(rsp), allocatable :: work(:)
+
+    m = size(A,dim=1)
+    n = size(A,dim=2)
+    mn = min(m,n)
+    lda = size(A,dim=1)
+    ldu = size(U,dim=1)
+    ldvt = size(Vt,dim=1)
+    @:ASSERT(all(shape(u) >= (/m,mn/)))
+    @:ASSERT(all(shape(vt) >= (/mn,n/)))
+    @:ASSERT(size(sigma) == mn)
+
+   lwork = max(1,3*min(m,n)+max(m,n),5*min(m,n))
+
+    allocate(work(lwork))
+
+    u = 0.0_rsp
+    vt = 0.0_rsp
+    sigma = 0.0_rsp
+    work = 0.0_rsp
+    info = 0
+
```

Comment:
Any reason to fill up those arrays with zeros? They are all intent(out) arguments for the following `sgesvd` call

Reply:
If unreferenced, would they be 0 for all compilers? If that is always safe, then it can be removed.

Reply:
No, if unreferenced their content would be arbitrary. But if they have the right size (see above) the F77-LAPACK routine should set all of them to some value, shouldn't it?

Reply:
Both u and vt might have unreferenced elements inside LAPACK if respective jobs are not 'A'. Sigma and work are harmless though, so will check in a change there.